### PR TITLE
Add config schema for password expiry

### PR DESCRIPTION
**Pull request by IniZio**
ref #3957

### CODE REVIEW

File: pkg/lib/authenticationflow/declarative/data_password_policy.go [hunk 1]
```diff
@@ -10,6 +10,15 @@ type PasswordPolicyHistory struct {
 	Days    int  `json:"days,omitempty"`
 }
 
+type PasswordPolicyExpiry struct {
```
Comment (louischan-oursky) [inline pkg/lib/authenticationflow/declarative/data_password_policy.go:13]: It seems to me that there is no need to tell password expiry to the authflow caller. The information we expose in authflow is for the caller to implement the UI. Password expiration details normally will not be shown in the UI. Most of the time, the end-user just enter their password and let the server to tell them whether the password is still valid.

So do not report password expiry in authflow data.

File: pkg/lib/config/authenticator.go [hunk 2]
```diff
@@ -58,34 +79,70 @@ var _ = Schema.Add("PasswordPolicyConfig", `
 		"minimum_guessable_level": { "type": "integer" },
 		"excluded_keywords": { "type": "array", "items": { "type": "string" } },
 		"history_size": { "type": "integer" },
-		"history_days": { "$ref": "#/$defs/DurationDays" }
+		"history_days": { "$ref": "#/$defs/DurationDays" },
+		"expiry": { "$ref": "#/$defs/PasswordExpiryConfig" }
```
Comment (louischan-oursky) [inline pkg/lib/config/authenticator.go:83]:
```yaml
password:
  # The original policy configuration.
  # I think password expiry should not be put in it.
  # The things in policy is more about password requirement.
  # If we can rename it, it should be called "requirement" instead.
  policy:
    min_length: 8
  # This force_change is for password.policy.
  force_change: true

  # The 3 cases can be turned on individually.
  # However, the precedence is deny_login > force_change > prompt_change.
  expiry: # "expiration" is American English, while "expiry" is British English. But we have been using "expiry" in the config. So let's stick with it.
    # In this case, the authflow will result in a dead end, with an error telling that the password is expired and the login is denied.
    deny_login:
      enabled: true
      since_last_update_in_days: 90
    # In this case, the authflow will enter the change_password step (The existing one). To proceed, the end-user must change the password.
    force_change:
      enabled: true
      since_last_update_in_days: 60
    # In this case, the authflow will enter the change_password step. But this step now supports taking an input to skip the password update.
    prompt_change:
      enabled: true
      since_last_update_in_days: 30
```

File: pkg/lib/config/authenticator.go [hunk 3]
```diff
@@ -86,6 +88,57 @@ func (c *PasswordPolicyConfig) SetDefaults() {
...
+}
+`)
+
+var _ = Schema.Add("PasswordExpiryForceChangeConfig", `
+{
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"enabled": { "type": "boolean" },
+		"since_last_update_in_days": { "$ref": "#/$defs/DurationDays" }
```
Comment (louischan-oursky) [inline pkg/lib/config/authenticator.go]: Use DurationString. Refer to RateLimitConfig to see how the schema of RateLimitConfig is written in a way that `since_last_update_duration` is only required when `enabled` is true.

File: pkg/lib/config/authenticator.go [hunk 4]
```diff
@@ -86,6 +88,57 @@ func (c *PasswordPolicyConfig) SetDefaults() {
...
+}
+
+func (c *PasswordExpiryConfig) SetDefaults() {
+	if c.ForceChange == nil {
+		c.ForceChange = &PasswordExpiryForceChangeConfig{}
+	}
+}
+
+type PasswordExpiryForceChangeConfig struct {
+	Enabled               bool         `json:"enabled"`
```
Comment (louischan-oursky) [inline pkg/lib/config/authenticator.go]: omitempty

File: pkg/lib/config/feature_authenticator.go [hunk 5]
```diff
@@ -65,3 +67,44 @@ func (c *PasswordPolicyItemFeatureConfig) SetDefaults() {
 		c.Disabled = newBool(false)
 	}
 }
+
+var _ = FeatureConfigSchema.Add("PasswordExpiryFeatureConfig", `
```
Comment (louischan-oursky) [inline pkg/lib/config/feature_authenticator.go:71]: No need to add feature config for password expiry.

File: pkg/lib/config/authenticator.go [hunk 6]
```diff
@@ -120,25 +120,15 @@ func (c *PasswordExpiryConfig) SetDefaults() {
...
-	Enabled               bool         `json:"enabled"`
-	SinceLastUpdateInDays DurationDays `json:"since_last_update_in_days,omitempty"`
+	Enabled         bool           `json:"enabled"`
+	SinceLastUpdate DurationString `json:"since_last_update,omitempty"`
 }
 
 func (c *PasswordExpiryForceChangeConfig) IsEnabled() bool {
-	return c.Enabled && c.SinceLastUpdateInDays > 0
+	sinceLastUpdate, _ := c.SinceLastUpdate.MaybeDuration()
+	return c.Enabled && sinceLastUpdate > 0
```
Comment (louischan-oursky) [inline pkg/lib/config/authenticator.go:129]: It should be `c.Enabled && ok && sinceLastUpdate > 0`.

**Comment by louischan-oursky** [comment]
Like pkg/lib/config/testdata/rate_limit_tests.yaml, add tests for the new config.

**Comment by IniZio** [comment]
It's currently added in `pkg/lib/config/testdata/config_tests.yaml`, or prefer splitting into separate test files?

Will add more test on the conditional require nevertheless

File: pkg/lib/config/authenticator.go (hunk 2 above)
Comment (louischan-oursky) [reply pkg/lib/config/authenticator.go]: - Rename `since_last_update` to `duration_since_last_update`.
- Copy the above config and paste it as comment around `PasswordExpiryConfig` for future reference.

File: pkg/lib/authn/authenticator/password/deps.go [hunk 7]
```diff
@@ -26,6 +26,10 @@ func ProvideChecker(
 		checker.PwHistoryDays = cfg.Policy.HistoryDays
 		checker.PasswordHistoryEnabled = cfg.Policy.IsEnabled()
 	}
+	if !*featureCfg.Password.Expiry.ForceChange.Disabled {
+		checker.PwExpiryForceChangeEnabled = cfg.Expiry.ForceChange.IsEnabled()
+		checker.PwExpiryForceChangeSinceLastUpdateInDays = cfg.Expiry.ForceChange.SinceLastUpdateInDays
+	}
```
Comment (louischan-oursky) [inline pkg/lib/authn/authenticator/password/deps.go]: Remove changes to authn/authenticator/password/checker.go and authn/authenticator/password/deps.go

**Comment by louischan-oursky** [comment]
superseded by https://github.com/authgear/authgear-server/pull/3968

**Pull request closed without merging**